Redpanda Self Test / Test Orchestration and Management - Part one

graphcareful opened the pull request:
## Cover letter

This is the first of a series of commits that make up the new self test feature, [RFC](https://docs.google.com/document/d/1UBc9CcD-mKkVOtPbO5G_t7UuU2GUIP3I0nnCPTDFpEY/edit#heading=h.fxgc4ycrks86)

This patch set contains the infrastructure and tooling necessary to start, stop and retrieve results from self tests. Clients control the behavior via the admin API. All brokers include a new self test RPC service used to communicate with each other. The leader node is responsible for test orchestration among all peers.

## Release notes
- none

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+
+    if (!_state.is_test_active()) {
+        /// test has finished
+        if (_state.current_uuid() == uuid_t{}) {
```
I don't fully understand the usage of a test id here. If all the tests are going to be identical, why not just have each local test result either be usable or not usable based on how recently the local test was run? Then we'd know whether a test result is valid or not solely based on whether it's stale, rather only being useful by exactly one global test.

andrwng replied:
Right I understand the mapping of local to global, but I wonder why that's desirable vs having local results being usable by multiple global tests. I suppose if a user wanted to run multiple tests in succession, the results will be more reasonable if we guaranteed a fresh run every time. Also if there are plans to extend the tests with different parameters having test ID seems important.

graphcareful replied:
Taking this advice testId will have much less of an importance in the next update, however i'm leaving the notion of test ID in, maybe it might be useful for the frontend to know if the system was out of sync or something like that.

graphcareful commented on src/v/cluster/self_test_backend.cc:
```diff
@@ -0,0 +1,114 @@
+}
+
+status_report self_test_backend::start_test(uuid_t id) {
+    auto units = _m.try_get_units();
+    if (units) {
+        _id = id;
+        _disk_test = disk_test();
+        _network_test = network_test();
+        (void)do_start_test(id)
+          .then([this](auto results) { _prev_run = std::move(results); })
+          .finally([units = std::move(units)] {});
+    }
+    return get_status_report();
```
`_id` is set as to an instance member variable and shipped to all nodes. Each node will return this given ID when queried about its status so we can discern if we are still in sync or not. The idea is to maintain loose coordination across start/stop calls without integration with `v/cluster`

For example if the test coordinator asks all nodes for their status, and retrieves non consistent test IDs that means that a job was already running, while a coordinator issued a new start command.

graphcareful replied:
This no longer applies, so resolving

graphcareful commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+
+    auto leader_opt = _leaders.local().get_leader(model::controller_ntp);
+    if (!leader_opt.has_value() || leader_opt != _self) {
+        // Only cluster leader can initiaite start_test() rpc calls
+        co_return self_test_status_report::failure;
```
Not all , some expect the calls to be issued to the leader. Good point though

graphcareful replied:
took this advice , resolving

graphcareful commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+ss::future<self_test_frontend::global_test_state> self_test_frontend::status() {
+    auto gate_holder = _gate.hold();
+
+    auto result = co_await invoke_on_all_nodes(
+      [this](model::node_id id, auto handle) {
+          return handle->query_status(!_state.has_completed(id));
+      });
```
I was thinking this too, this is why i had the concept of `participants`. The idea is to ensure logical consistency across all `participents` for the duration of the tests execution. So if a new `participent` starts returning that it is doing work once it had already expected to be completed, we can detect that here.

graphcareful replied:
I think it cannot be made const because of the RPC client, i.e. the instance methods of generated client code are not const

graphcareful replied:
FWIW I'm considering modifying this to not sending the RPCs if they aren't needed

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +168,65 @@
+                    "type": "long",
+                    "description": "node id"
+                },
+                "status": {
+                    "type": "long",
+                    "description": "Running or idle status"
+                },
```
+1 JSON APIs should use stringized enum names

jcsp commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+    co_return enrich(node_id(), v);
+}
+
+ss::future<self_test_frontend::local_test_state>
```
Yep, "node" sounds like the right noun to me.

jcsp replied:
:point_up: did this comment got missed in revisions?

I don't mind if it's addressed in a followup, just making sure it isn't lost.

graphcareful replied:
Nope i forgot this one, thanks!

jcsp commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,93 @@
+}
+
+struct start_test_request
+  : serde::envelope<start_test_request, serde::version<0>> {
```
This'll need an update when you rebase, to use an explicit compat_version

jcsp commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,93 @@
+};
+
+struct test_results : serde::envelope<test_results, serde::version<0>> {
+    long disk_test_stoptime{0};
```
Why a `long` rather than a time type?  If we don't have serialization for time types should probably add it, rather than using generic integers

jcsp commented on src/v/cluster/self_test_frontend.h:
```diff
@@ -0,0 +1,190 @@
+    global_test_state previous_run() const { return _previous_successful_run; }
+
+private:
+    class invoke_wrapper {
```
Please add a comment explaining the purpose of the class

jcsp commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+    const auto st = co_await status();
+    if (st.is_out_of_sync()) {
+        // Result contains inconsistencies, cancel and report why
+        co_await stop_test();
```
Once we've stopped the test, the system is back in a consistent state and we could proceed to start the test as requested

jcsp commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,305 @@
+
+    /// Invoke command to start test on all nodes, using the same test id
+    const auto id = uuid_t::create();
+    const auto results = co_await invoke_on_all_nodes(
```
Bit unclear whether the "results" here are an interim status report, or if the start_test RPC blocks until the test is complete?

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -40,6 +40,63 @@
+                            "schema": "json"
+                        },
+                        "500": {
+                            "description": "Test failed to start",
```
500 means something totally unexpected: if the test cannot start right now for known reasons, that would be a 503, whereas if we get a 500 we cannot say anything about why (it's not something we anticipated)

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +168,65 @@
+            "id": "self_test_report",
+            "description": "Self test single test report",
+            "properties": {
+                "disk_test_stoptime": {
```
I'm confused about stoptime -- elsewhere in the code it looks like the test is intentionally waiting for stoptime (like it's a target runtime), but here it seems to be described as part of the report like it's a variable runtime

graphcareful replied:
Yeah this is stub code to be removed and replaced with the actual result types from the respective disk + network benchmarks

graphcareful replied:
I should/will document this in the property description here

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +168,65 @@
+                },
+                "failures": {
+                    "type": "boolean",
+                    "description": "True if any failures occurred"
```
Would be more helpful for the user if we use a string (or vector of strings) for errors, rather than a bool (if this is true they presumably have to spelunk logs to find out why?)

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +168,65 @@
+        },
+        "self_test_node_status": {
+            "id": "self_test_node_status",
+            "description": "Self test status",
```
"report" and "status" are pretty much synonyms, in these two  API endpoint.  Maybe "result" is a better noun to express the idea of something that comes out at the end, vs. something that is mutable and will change as time goes on.

Or... do we need two endpoints?  The status endpoint includes the report anyway.

graphcareful replied:
type removed, resolving

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +168,65 @@
+        },
+        "self_test_response": {
+            "id": "self_test_response",
+            "description": "Self test response",
```
This description is a bit nondescript :-)

It's a response, but from what?  Should it be self_test_status_response or something?

graphcareful replied:
type removed

jcsp commented on src/v/redpanda/admin_server.cc:
```diff
@@ -2898,6 +2899,79 @@ void admin_server::register_transaction_routes() {
 }
 
+void admin_server::register_self_test_routes() {
+    register_route<user>(
```
In the rest of the API, the pattern is generally `user` for read methods, and `superuser` for endpoints that change something.

jcsp commented on src/v/redpanda/admin_server.cc:
```diff
@@ -2898,6 +2899,79 @@ void admin_server::register_transaction_routes() {
+            });
+
+          using result_t = ss::httpd::debug_json::self_test_node_status;
+          ss::httpd::debug_json::self_test_response r;
```
nit: could reserve the r.reports vector size ahead of loop

graphcareful replied:
Hmm seems like the json::json_list object doesn't have a reserve method

jcsp commented on src/v/redpanda/admin_server.cc:
```diff
@@ -2898,6 +2899,79 @@ void admin_server::register_transaction_routes() {
+              ns.node_id = lts.node_id;
+              ns.status = std::underlying_type_t<cluster::self_test_status>(
+                lts.status);
+              ns.was_available = lts.is_available;
```
Bit odd to alias from `is` to `was`?

jcsp reviewed (commented):
No major issues, bunch of small comments -- looks like a sound basis for following work.

graphcareful commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,93 @@
+
+namespace cluster {
+
+enum class self_test_status : int8_t { running = 0, idle, cancelled };
```
I really didn't like this name, i'm terrible at naming though :/ , I'll give it another shot, let me know what you think

graphcareful commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,93 @@
+struct stop_test_request
+  : serde::envelope<stop_test_request, serde::version<0>> {
+    using rpc_adl_exempt = std::true_type;
+    int8_t empty;
```
Placeholder as the `input` and `output` types for an RPC endpoint must be a well defined struct. The situation is `stop_test()` doesn't take any inputs.

In either case, I just checked if its necessary to have any fields in a struct you desire to use as an RPC input, and its actually not necessary. Going to make a proper `empty_request` type that contains 0 member fields.

graphcareful commented on src/v/cluster/self_test_frontend.h:
```diff
@@ -0,0 +1,190 @@
+
+namespace cluster {
+
+enum class self_test_status_report : int8_t {
```
removed

graphcareful commented:
Much has changed in the latest series of force-pushes:
1. Nodes that have already reported success will not be periodically queried for status, as soon as a node returns results, its removed from the set of IPs to query the next time a status() request is made. This is repeated until all participants report they are running no self test jobs.
2. Notion of 'unreachable' node introduced in `self_test_status`. Retry policy of 5 attempts is introduced as well. After the fifth try the node will be removed from the list of participants and included in the result set as an unreachable node.
3. Test uuid is less significant. All checks for out of sync situations etc removed. Each broker just reports uuid and results to client, client can do what it wants with that information.
4. Full 3 node ducktape test included - includes test of stop_test() endpoint.
5. The test_stub updated to have its `cancel()` method not return a future. Therefore stop_test() sets a conditional and waits for the general _mutex to be available before it returns. Even though this stub is to be deleted it was updated to reflect how the real self_tests will operate.
6. Removed the general map<string,string> that represented test options / test results and replaced with actual types.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=6948) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=6948) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: graphcareful<br/>:x: Rob Blafford<br/><hr/>**Rob Blafford** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=6948) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=6948) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/redpanda-data/redpanda?pullRequest=6948) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: graphcareful<br/>:x: Rob Blafford<br/><hr/>**Rob Blafford** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/redpanda-data/redpanda?pullRequest=6948) it.</sub>

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +167,64 @@
+            "properties": {
+                "test_duration": {
+                    "type": "long",
+                    "description": "Length of time the test took to complete"
```
Since this an integer, description should say what unit the value is in (presumably seconds).

Can pick this up later when adding the real checks if you like.

graphcareful replied:
Yeah this is going away, stub code for now

jcsp commented on src/v/redpanda/admin/api-doc/debug.json:
```diff
@@ -111,6 +167,64 @@
+                },
+                "status": {
+                    "type": "long",
+                    "description": "One of either idle / running / unreachable"
```
Description  gives strings but value is integer

jcsp commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+    start_test_request req{.id = id, .dto = dto, .nto = nto};
+    co_await invoke_on_all_nodes(
+      [req](auto handle) { return handle->start_test(req); });
+    /// TODO: Throw if something bad occurred
```
Is the comment out of date?  Since we co_await on the invoke_on_all_nodes, hopefully we're already throwing if something went wrong.

jcsp reviewed (approved):
Thanks for the updates, just minor notes.  @graphcareful since this is a staged thing, I leave it up to you which comments you want to address here vs. in the next PR

andrwng commented on src/v/cluster/self_test_frontend.h:
```diff
@@ -0,0 +1,169 @@
+
+    ss::sharded<cluster::members_table>& _members;
+    ss::sharded<cluster::self_test_backend>&
+      _self_be; // only invoked to execute the self test on 'this' node, the
+                // leader node where this frontend will be active as well
```
nit: seems like the status quo is to comment the line above the definition?

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+    auto brokers = get_active_participants();
+    auto local_state_vec = co_await ssx::parallel_transform(
+      brokers, [this, f = std::move(f)](const auto& node_id) {
+          ss::shared_ptr<invoke_wrapper> handle;
```
nit: I'm curious, maybe this will be more useful in later PRs, but if this is owned and passed around by this fiber, would it make sense for it to just be a `unique_ptr`?

graphcareful replied:
nice catch, was leftover from prev impl

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+    if (!_state.finished()) {
+        co_return _state;
+    }
+    _previous_successful_run = _state;
```
nit: maybe call `_state` `_pending_run` or something? Then it's a bit clearer what it's for and that it's type is the same as `_previous_successful_run`

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+std::vector<model::node_id>
+self_test_frontend::get_active_participants() const {
+    auto participants = _state.active_participant_ids();
+    auto brokers = participants.empty() ? _members.local().node_ids()
```
This seems a bit nuanced and probably deserves a comment. If there are no more pending participants, that means the previous test is done, right? Which is why we return all nodes? I find it a bit confusing because I think there's an assumption that after a test finishes, the next call will be `start_test`, otherwise we end up sending RPCs fruitlessly for a test that didn't finish?

graphcareful replied:
Return all nodes because its assumed a new test run is to start. Its up to the caller to check the current status of the world before calling the invoke_on_all method.

graphcareful replied:
commenting and moving into anon-lambda to remove confusion

andrwng commented on src/v/cluster/self_test_frontend.h:
```diff
@@ -0,0 +1,169 @@
+    ss::future<global_test_state> stop_test();
+    ss::future<global_test_state> status();
+
+    global_test_state previous_run() const { return _previous_successful_run; }
```
Unused?

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+    }
+    _previous_successful_run = _state;
+    _state = global_test_state({});
+    co_return _previous_successful_run;
```
Maybe this will be used in the future, but if not, it seems like this is never used and we might as well not bother keeping track of it?

graphcareful replied:
Currently status asks all nodes their status in the case things are finished instead of returning cached result. Wasn't intended, going to return this instead.

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,263 @@
+
+ss::future<self_test_frontend::global_test_state>
+self_test_frontend::stop_test() {
+    auto gate_holder = _gate.hold();
+    int8_t retries = 3;
+    auto test_state = co_await status();
```
I get the feeling we'd gain a lot of code clarity by explicitly keeping track of the state of the most recently run test. For instance, a `status()` call would be able to immediately return if the test were already finished rather than sending out RPCs; likewise, a `stop_test()` call on a test that has already completed would avoid similar work.

graphcareful replied:
Ah i remember why I didn't do that now, because in the case leadership changes then previous results wouldn't be cached on new leader. Asking nodes for their results via one RPC to each isn't such a big deal.

graphcareful replied:
I can probably do something a little more clever like checking if previous results is null or not before making that call, changes incoming..

graphcareful commented on src/v/redpanda/admin_server.cc:
```diff
@@ -2929,6 +2930,124 @@ void admin_server::register_transaction_routes() {
+      cluster::self_test_frontend::shard,
+      [](auto& self_test_frontend) { return self_test_frontend.stop_test(); });
+    if (!r.finished()) {
+        throw ss::httpd::base_exception(
```
yeah, admin server logs all exception though so this will be logged at error

graphcareful commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,178 @@
+struct diskcheck_result
+  : serde::
+      envelope<diskcheck_result, serde::version<0>, serde::compat_version<0>> {
+    std::optional<ss::sstring> warnings;
+    std::optional<ss::sstring> errors;
```
As it stands now they just return the stringified exception to the user, so i'll just rename them to not be plural

graphcareful commented on src/v/cluster/self_test_rpc_types.h:
```diff
@@ -0,0 +1,178 @@
+    using rpc_adl_exempt = std::true_type;
+
+    uuid_t id;
+    diskcheck_opts dto;
+    netcheck_opts nto;
```
agreed, saving that for later work

graphcareful commented:
Final last major changes:
1. New result type that represents a generic response from any test, includes latency throughput fields
2. swagger modified to return this new type as a result from a test
3. Name changes and smaller nits suggested in last review implemented
4. Suggestion to optionally run one or more or all of the self tests implemented

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,266 @@
+        return participants.empty() ? _members.local().node_ids()
+                                    : participants;
+    }();
+    auto local_state_vec = co_await ssx::parallel_transform(
```
nit: `node_state_vec`, to follow the rename?

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,266 @@
+            });
+      });
+    auto result = global_test_state(std::move(local_state_vec));
+
+    /// Condense response from all nodes to new known state of the system
+    _pending_run.merge(std::move(result));
+
+    if (!_pending_run.finished()) {
+        co_return _pending_run;
+    }
+    auto previous_successful_run = _pending_run;
+    _pending_run = global_test_state({});
+    co_return previous_successful_run;
```
I'm liking how we're getting to be more stateless now that we're no longer keeping the previous run around. Could we take this even further by not tracking `_pending_run` at all, and instead always use the full set of node IDs returned by the members table?

As implemented, we're saving on some duplicate RPCs, but we also discussed that the RPCs were pretty lightweight given the caching within the test workers. I think it'd be easier to reason about without knowing the nitty-gritty of this implementation, though maybe I'm missing something

graphcareful replied:
Recently removed as per offline discussion

andrwng commented on tests/rptest/tests/self_test_test.py:
```diff
@@ -0,0 +1,96 @@
+from ducktape.utils.util import wait_until
+
+
+class SelfTestTest(RedpandaTest):
```
It'd be great to add some negative tests. Perhaps a test for what happens if controller leadership changes mid-global-test? And a test where some workers crash mid-local-test?

graphcareful replied:
done

andrwng replied:
Would still be great if we could add a test that exercises a leadership change, ensuring that the results from on-going tests are still honored by the new leader.

Feel free to pick that up as in a separate PR though

andrwng commented on src/v/cluster/self_test_frontend.h:
```diff
@@ -0,0 +1,176 @@
+      : std::runtime_error(std::move(msg)) {}
+};
+
+class self_test_frontend {
```
nit: here might be a good place to add a high level overview of how that self-test infra comes together -- that requests are meant to be sent to the leader, and they RPC out to all nodes, etc. Also that it's not critical to be done on the leader only (or maybe it is so our members table is most up-to-date?)

andrwng reviewed (commented):
Looking a lot cleaner! I still have question about the state maintained for the global test, and a couple test suggestions (though maybe those are better suited for follow-up PRs?)

andrwng commented on src/v/cluster/self_test_frontend.cc:
```diff
@@ -0,0 +1,205 @@
+        if (!test_state.finished()) {
+            vlog(
+              clusterlog.warn,
+              "Failed to sucessfully stop all outstanding jobs on all "
+              "participants: {} , trying again, retries: {}",
+              test_state.active_participant_ids(),
+              retries);
```
nit: this is a little confusing since we're not actually showing all the participants. How about something like

"Failed to stop self tests; remaining running tests at nodes {active_participant_ids}, remaining retries {retries}"

graphcareful replied:
I decided to print only the participants for which contain the stuck jobs

andrwng replied:
Right, I just think it's worth updating the wording to make it clear that the printed participants are the _remaining_ ones, not _all_ participants

andrwng commented on src/v/redpanda/admin_server.cc:
```diff
@@ -2929,6 +2930,130 @@ void admin_server::register_transaction_routes() {
+ss::future<ss::json::json_return_type>
+admin_server::self_test_get_results_handler(
+  std::unique_ptr<ss::httpd::request>) {
+    namespace dbg_ns = ss::httpd::debug_json;
```
nit: for posterity I think it's worth adding a comment mentioning/explaining why we're explicitly not redirecting to the leader for this operation (it's read-only and inexpensive to send anywhere). It's otherwise tempting to look at this and assume it was an oversight

andrwng reviewed (approved):
Liking this new structure. Just a couple of remaining nits and a test suggestion, but feel free to pick them up in a separate PR if this is blocking other changes
